[Audio Editor] Port Audio Options from SoH Audio Editor

Glought opened the pull request:
I have port over Audio Options from SoH Audio Editor.

- Moved "Mute Low HP Alarm" to Audio Editor -> Audio Options
   Changed Cvar
   From: "gEnhancements.Sfx.LowHpAlarm"
   To: "gAudioEditor.LowHpAlarm"

- Moved "Mute Carpenter Sounds" to Audio Editor -> Audio Options
   Changed Cvar
   From: "gEnhancements.Sfx.MuteCarpenterSfx"
   To: "gAudioEditor.MuteCarpenterSfx"

- Moved "Mute Crying Goron Child" to Audio Editor -> Audio Options
  Changed Cvar
  From: "gEnhancements.Sfx.ChildGoronCry"
  To: "gAudioEditor.ChildGoronCry"

- Added "Disable Tatle Call Audio".
- Added "Disable Enemy Proximity Music"  -> #1302.
- Added "Randomize All Music and Sound Effects on New Scene".
- Added "Randomize All Music and Sound Effects on Randomizer Generation".
- Added "Display Sequence Name on Overlay" checkbox  and "Overlay Duration" Slider.
- Added "Enable Link's Voice Pitch Multiplier" checkbox and "Link's Voice Pitch Multiplier"  Slider.
- Added "AddSearchWidget" Functionally .

<img width="891" height="757" alt="2ship audio editor - ported 2" src="https://github.com/user-attachments/assets/56ca404b-c762-49f0-8ca6-b58cc9bfba45" />


<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4608882245.zip)
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4608936959.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4609154782.zip)
<!--- section:artifacts:end -->

Glought committed:
Ported Audio Options from SoH Audio Editor

*Moved "Mute Low HP Alarm" to Audio Editor -> Audio Options
   Changed Cvar
   From: "gEnhancements.Sfx.LowHpAlarm"
   To: "gAudioEditor.LowHpAlarm"
*Moved "Mute Carpenter Sounds" to Audio Editor -> Audio Options
   Changed Cvar
   From: "gEnhancements.Sfx.MuteCarpenterSfx"
   To: "gAudioEditor.MuteCarpenterSfx"
*Moved "Mute Crying Goron Child" to Audio Editor -> Audio Options
   Changed Cvar
   From: "gEnhancements.Sfx.ChildGoronCry"
   To: "gAudioEditor.ChildGoronCry"
*Added "Disable Tatle Call Audio"
*Added "Disable Enemy Proximity Music"
*Added "Randomize All Music and Sound Effects on New Scene"
*Added "Randomize All Music and Sound Effects on Randomizer Generation"
*Added "Display Sequence Name on Overlay" and "Overlay Duration" Slider
*Added "Link's Voice Pitch Multiplier" Slider

Glought committed:
*Ported "AddSearchWidget" Functionally
  Enabled for Audio Editor -> Audio Options

Glought committed:
Cleaned up some code I missed and ran clang format.

Glought committed:
Renamed "OnScene" Hook to "OnSceneChange"

Glought committed:
Made it so "Link's Voice Pitch Multiplier" slider only effects Link's voice.

Glought committed:
Forgot the "AudioHook.cpp" file its needed for the
"Display Sequence Name on Overlay" option.

Glought committed:
Moved the "Sfx" folder to the "Audio" folder.

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -10,13 +10,33 @@
 #include <ship/utils/StringHelper.h>
 #include "../../BenGui/UIWidgets.hpp"
+#include "../../BenGui/BenMenu.h"
+#include "../../BenGui/BenGui.hpp"
```
```suggestion
#include "2s2h/BenGui/BenMenu.h"
#include "2s2h/BenGui/BenGui.hpp"
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -416,17 +436,25 @@ void DrawTypeChip(SeqType type) {
-    //});
+void AudioEditorRegisterOnSceneChangeHook() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnSceneChange>([](int16_t sceneNum) {
+        if (CVarGetInteger("gAudioEditor.RandomizeAllOnNewScene", 0)) {
```
The pattern in this file is to use `CVAR_AUDIO(var)`, so for consistency's sake, let's change these CVar calls.


```suggestion
        if (CVarGetInteger(CVAR_AUDIO("RandomizeAllOnNewScene"), 0)) {
```

I realize this may be moot when it comes to addressing CVar prefixes in general, but within the scope of this PR, I'd prefer we not have "gAudioEditor.Something" 10 times in one file.

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -416,17 +436,25 @@ void DrawTypeChip(SeqType type) {
+
+void AudioEditorRegisterOnGenerationCompletionHook() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnGenerationCompletion>([]() {
+        if (CVarGetInteger("gAudioEditor.RandomizeAllOnRandoGen", 0)) {
```
```suggestion
        if (CVarGetInteger(CVAR_AUDIO("RandomizeAllOnRandoGen"), 0)) {
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    lowHpAlarm = { .name = "Mute Low HP Alarm", .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    lowHpAlarm.CVar("gAudioEditor.LowHpAlarm")
```
```suggestion
    lowHpAlarm.CVar(CVAR_AUDIO("LowHpAlarm"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    muteCarpenterSfx = { .name = "Mute Carpenter Sounds", .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    muteCarpenterSfx.CVar("gAudioEditor.MuteCarpenterSfx")
```
```suggestion
    muteCarpenterSfx.CVar(CVAR_AUDIO("MuteCarpenterSfx"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    childGoronCry = { .name = "Mute Crying Goron Child", .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    childGoronCry.CVar("gAudioEditor.ChildGoronCry")
```
```suggestion
    childGoronCry.CVar(CVAR_AUDIO("ChildGoronCry"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    tatlCall = { .name = "Disable Tatl Call Audio", .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    tatlCall.CVar("gAudioEditor.DisableTatlCallAudio")
```
```suggestion
    tatlCall.CVar(CVAR_AUDIO("DisableTatlCallAudio"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    enemyProx = { .name = "Disable Enemy Proximity Music", .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    enemyProx.CVar("gAudioEditor.EnemyBGMDisable")
```
```suggestion
    enemyProx.CVar(CVAR_AUDIO("EnemyBGMDisable"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+    randoMusicOnSceneChange = { .name = "Randomize All Music and Sound Effects on New Scene",
+                                .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    randoMusicOnSceneChange.CVar("gAudioEditor.RandomizeAllOnNewScene")
```
```suggestion
    randoMusicOnSceneChange.CVar(CVAR_AUDIO("RandomizeAllOnNewScene"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+    randomAudioOnSeedGen = { .name = "Randomize All Music and Sound Effects on Randomizer Generation",
+                             .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    randomAudioOnSeedGen.CVar("gAudioEditor.RandomizeAllOnRandoGen")
```
```suggestion
    randomAudioOnSeedGen.CVar(CVAR_AUDIO("RandomizeAllOnRandoGen"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    displaySeqName = { .name = "Display Sequence Name on Overlay", .type = WidgetType::WIDGET_CVAR_CHECKBOX };
+
+    displaySeqName.CVar("gAudioEditor.SeqNameNotification")
```
```suggestion
    displaySeqName.CVar(CVAR_AUDIO("SeqNameNotification"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    ovlDuration = { .name = "Overlay Duration: %d seconds", .type = WidgetType::WIDGET_CVAR_SLIDER_INT };
+
+    ovlDuration.CVar("gAudioEditor.SeqNameNotificationDuration")
```
```suggestion
    ovlDuration.CVar(CVAR_AUDIO("SeqNameNotificationDuration"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+
+    voicePitch = { .name = "Link's Voice Pitch Multiplier", .type = WidgetType::WIDGET_CVAR_SLIDER_FLOAT };
+
+    voicePitch.CVar("gAudioEditor.LinkVoiceFreqMultiplier")
```
```suggestion
    voicePitch.CVar(CVAR_AUDIO("LinkVoiceFreqMultiplier"))
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
+                     .Color(THEME_COLOR)
+                     .Tooltip("Mutes the beeping alarm when you are critically low on health."));
+
+    BenGui::mBenMenu->AddSearchWidget({ lowHpAlarm, "Enhancements", "Audio Editor", "Audio Options" });
```
With this being identical for the 10 different calls aside from the widget argument itself maybe we could make this a function call? Something like:

```c++
void AddAudioSearchWidget(WidgetInfo& widgetInfo) {
    BenGui::mBenMenu->AddSearchWidget({ widgetInfo, "Enhancements", "Audio Editor", "Audio Options" });
}
```

And then replace each use you have here with:

```c++
AddAudioSearchWidget(lowHpAlarm);
```

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -760,4 +746,104 @@ void AudioEditor_UnlockAll() {
\ No newline at end of file
+}
+
+void RegisterAudioWidgets() {
```
Can you remove the extra line breaks within each section? For example, the `lowHpAlarm` section should have no blank lines within it, but it should have one blank line after to separate it from the next `muteCarpenterSfx` section. I think it would be easier to tell at a glance how everything is grouped.

Eblo commented on mm/2s2h/GameInteractor/GameInteractor_HookTable.h:
```diff
@@ -52,3 +53,9 @@ DEFINE_HOOK(OnItemGive, (u8 item))
 DEFINE_HOOK(ShouldVanillaBehavior, (GIVanillaBehavior flag, bool* should, va_list originalArgs))
+
+// Audio
+DEFINE_HOOK(OnSeqPlayerInit, (int32_t playerIdx, int32_t seqId));
```
```suggestion
DEFINE_HOOK(OnSeqPlayerInit, (s32 playerIdx, s32 seqId));
```

Eblo commented on mm/2s2h/GameInteractor/GameInteractor_HookTable.h:
```diff
@@ -52,3 +53,9 @@ DEFINE_HOOK(OnItemGive, (u8 item))
+DEFINE_HOOK(OnSeqPlayerInit, (int32_t playerIdx, int32_t seqId));
+
+// Rando
+DEFINE_HOOK(OnGenerationCompletion, ());
```
While the comment is there, I think the hook name should also clarify that this is rando-specific. Maybe `OnRandoSeedGeneration`? Generally our "On" hooks happen after something, so "completion" is not needed in the name.

Eblo commented on mm/2s2h/GameInteractor/GameInteractor.h:
```diff
@@ -726,6 +726,7 @@ void GameInteractor_ExecuteAfterInterfaceClockDraw();
 void GameInteractor_ExecuteBeforeInterfaceClockDraw();
 
 void GameInteractor_ExecuteOnSceneInit(s16 sceneId, s8 spawnNum);
+void GameInteractor_ExecuteOnSceneChange(s16 sceneId);
```
What about `OnSceneInit` was insufficient to warrant this new hook?

Glought replied:
When I tried using `OnSceneInit` the music stopped playing. From my testing the music does get randomized but 
in the file `"mm/src/code/z_play.c" line 2497:  "GameInteractor_ExecuteOnSceneInit(sceneIdAbsolute, spawnNum);"` causes the music to stop playing. When I commented it out music will play. I don't know the exact purpose of that line I didn't want to risk breaking something by changing or removing it so I made `OnSceneChange` if you have any suggestions please let me know.

garrettjoecox replied:
Could you also try `AfterRoomSceneCommands` and `OnRoomInit`? If we do need to go with this place, I'll do some digging and figure out what the right name is, I don't think SceneChange is it

Eblo replied:
> causes the music to stop playing. When I commented it out music will play.

It may or may not be relevant, but there is a BGM replay patch that kills the main sequence player on transitions if certain settings are active. See `VB_PLAY_SCENE_SEQUENCE`. Try using `OnSceneInit` with that setting disabled if you haven't already.

Glought replied:
I never had that setting enabled when I tried using `OnSceneInit.`  
I just tested using `OnRoomInit` it does work and I don't need to add `GameInteractor_ExecuteOnRoomInit` to `z_play2SH.cpp: OTRPlay_SpawnScene` function.

Glought replied:
I pushed a commit removing `OnSceneChange` and to use 'OnRoomInit' instead.

Eblo commented on mm/src/code/z_actor.c:
```diff
@@ -2485,7 +2488,17 @@ void func_800B8E1C(PlayState* play, Actor* actor, f32 arg2, s16 arg3, f32 arg4)
+        }
+
+        AudioSfx_PlaySfx(sfxId, &player->actor.projectedPos, 4, &freqMultiplier, &gSfxDefaultFreqAndVolScale,
+                         &gSfxDefaultReverb);
```
I'm not sure I like introducing a new static variable, condition flow, and inline CVar into source code for this one. I think the better route would be to wrap the normal `AudioSfx_PlaySfx` call below in a GameInteractor_Should call with a new VB, then put the custom sfx stuff in an enhancement file.

Eblo commented on mm/src/code/z_parameter.c:
```diff
@@ -5091,12 +5091,15 @@ void Interface_SetTatlCall(PlayState* play, u16 tatlCallState) {
-        }
-        if (tatlCallState == TATL_STATE_2A) {
-            Audio_PlaySfx_AtPosWithReverb(&gSfxDefaultPos, NA_SE_VO_NA_HELLO_2, 0x20);
+        if (!CVarGetInteger("gAudioEditor.DisableTatlCallAudio", 0)) {
```
Ditto, I'd rather this be a VB instead.

Eblo commented on mm/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12207,7 +12207,7 @@ void Player_UpdateCamAndSeqModes(PlayState* play, Player* this) {
         }
 
-        if (play->actorCtx.attention.bgmEnemy != NULL) {
+        if (play->actorCtx.attention.bgmEnemy != NULL && !CVarGetInteger("gAudioEditor.EnemyBGMDisable", 0)) {
```
Ditto again, VB.

Glought committed:
Removed extra line breaks between sections in "RegisterAudioWidgets()".

Glought committed:
Renamed "OnGenerationCompletion" hook to "OnRandoSeedGeneration"

Glought committed:
Changed "gAudioEditor.*" to use "CVAR_AUDIO("*") in "AudioEditor.cpp"

Glought committed:
Added "AddAudioSearchWidget" function
Changed all "BenGui::mBenMenu->AddSearchWidget lines to use
"AddAudioSearchWidget".

Glought committed:
Changed the "OnSeqPlayerInit" hook parameters type to use the "s32" Macro.

Glought committed:
Update BenMenu/BenGui include paths: Replaced "../../" with "2s2h/".

Glought committed:
Ran Clang Format.

Glought committed:
Added "DisableEnemyProximityMusic.cpp"
Added "VB_PLAY_ENEMY_PROXIMITY_MUSIC" enum to "GIVanillaBehavior" in
"GameInteractor.h"
"Audio_UpdateEnemyBgmVolume" is now wrapped in a "GameInteractor_Should"

Glought committed:
Added "DisableTatlCallAudio.cpp"
Added "VB_PLAY_TATL_CALL_AUDIO" enum to "GIVanillaBehavior" in
"GameInteractor.h"
"Audio_PlaySfx(NA_SE_VO_NAVY_CALL)" is now wrapped in a "GameInteractor_Should"

Glought committed:
Added "LinksVoicePitchMultiplier.cpp"
Added "Enable Link's Voice Pitch Multiplier" checkbox to Audio Editor
The "Link's Voice Pitch Multiplier" slider is now disabled when
"Enable Link's Voice Pitch Multiplier" is not enabled.
Added "VB_LINK_VOICE_PITCH_MULTIPLIER" enum to "GIVanillaBehavior" in
"GameInteractor.h"
"AudioSfx_PlaySfx" in the file "z_actor.c" in function "Player_PlaySfx"
is now wrapped in a "GameInteractor_Should".

Caladius commented:
Just wanted to say thank you for taking the time and effort to port some of these things over to 2Ship.

garrettjoecox commented on mm/2s2h/BenGui/MenuTypes.h:
```diff
@@ -253,6 +260,15 @@ struct SidebarEntry {
     std::vector<std::vector<WidgetInfo>> columnWidgets;
 };
 
+struct SearchWidget {
```
Is this also something ported from SoH? Is the intent that we can have searchable widgets that may not be apart of a more formal "modern menu" setup?

Glought replied:
Yes its ported from SoH it's used by `void AddSearchWidget(SearchWidget widget);` in Menu.h. Without it the options in AudioEditor are unsearchable.

The PR on Shipwright is ` Search Exclusion and Search Widgets #5656 `  I only ported what was needed  for `AddSearchWidget` to work.

Glought committed:
Removed "OnSceneChange" Hook.
Removed "GameInteractor_ExecuteOnSceneChange".
Removed "GameInteractor_ExecuteOnSceneChange(play->sceneId);" from
"z_play_2SH.cpp:OTRPlay_SpawnScene" function
Renamed "AudioEditorRegisterOnSceneChangeHook" to
AudioEditorRegisterOnRoomInitHook".

Glought commented:
I have updated the PR message to reflect the changes I made.

Eblo commented on mm/2s2h/Enhancements/Audio/AudioEditor.cpp:
```diff
@@ -416,17 +437,25 @@ void DrawTypeChip(SeqType type) {
-    //    }
-    //});
+void AudioEditorRegisterOnRoomInitHook() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnRoomInit>([](s8 spawnNum, s8 roomNum) {
```
With `OnRoomInit`, the audio gets shuffled every time the player enters a new *room* rather than a scene (think anytime you use a door, or traverse the Woods of Mystery). I'm not sure if that was your intent, but it feels rather overwhelming to me.

Glought replied:
On further testing with `OnSceneInit` it semi works it doesn't play music in Clock Tower, Clock Town, Stock Pot Inn, Termina Field, Mayors Office, Great Bay, but will play music in the Milk bar, Honey & Darling's shop, Shops.

Eblo replied:
Just to be sure, have you tried it with all of these settings off?

- Faster scene transitions
- Debug mode
- Pause save

The BGM replay thing applies if any one of those is enabled.

Glought replied:
still doesn't work with those settings off.

Eblo replied:
Just for the heck of it, since the VB already exists, why not try your `AudioEditor_RandomizeAll` call within a     `COND_VB_SHOULD(VB_PLAY_SCENE_SEQUENCE...`? That runs just before a scene decides to play its music. I'm not 100% sure it will work, but might as well try.

Glought replied:
I even tested by disabling BGM Replay in code it still doesn't work.

Glought replied:
Calling `AudioEditor_RandomizeAll ` within the `COND_VB_SHOULD` worked.

Glought replied:
Some things I notice it doesn't get triggered when moving to other parts of clock town, termina Field. It triggers when entering shops,inn,etc,  or when leaving clock town  `clock town -> termina field`  or 
`Termina field -> Woodfall`. When calling `AudioEditor_RandomizeAll ` outside the `COND_VB_SHOULD ` it would randomize every time the scene changed.

Eblo replied:
That is curious. What exactly does your condition look like, and what do you mean by calling the randomization outside? I know sometimes music persists between scenes like in Clock Town, but that condition is evaluated after the placement of `VB_PLAY_SCENE_SEQUENCE`.

Glought replied:
```c++
void AudioEditorRegisterOnSceneInitHook() {
    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnSceneInit>([](s8 sceneId, s8 spawnNum) {
        COND_VB_SHOULD(VB_PLAY_SCENE_SEQUENCE, CVAR_AUDIO("RandomizeAllOnNewScene"), {
            AudioEditor_RandomizeAll();
        });
    });
}
```

Eblo replied:
A `COND_VB_SHOULD`, as well as any other hook macro, should be called at the top level of a register func, not within another hook. So when you're calling it "outside" in this case, that's effectively no different than just directly doing it in the OnSceneInit hook. Try this, remembering to call the register func itself in the overall InitElement:

```c++
void AudioEditorRegisterPlaySceneSequenceHook() {
    COND_VB_SHOULD(VB_PLAY_SCENE_SEQUENCE, CVAR_AUDIO("RandomizeAllOnNewScene"), {
        AudioEditor_RandomizeAll();
    });
}
```

Glought replied:
What about this
```c++
void AudioEditorRegisterRandomizeAllOnNewSceneHook() {
    COND_VB_SHOULD(VB_RANDOMIZE_ALL_ON_NEW_SCENE, CVarGetInteger(CVAR_AUDIO("RandomizeAllOnNewScene"), 0), {
        if (CVarGetInteger(CVAR_AUDIO("RandomizeAllOnNewScene"), 0)) {
            AudioEditor_RandomizeAll();
        }
    });
}
``` 
and calling it at the end of  `z_play_2SH.cpp`  in the `OTRPlay_SpawnScene` function.

```c++
extern "C" void OTRPlay_SpawnScene(PlayState* play, s32 sceneId, s32 spawn) {
    s32 pad;
    SceneTableEntry* scene = &gSceneTable[sceneId];

    scene->unk_D = 0;
    play->loadedScene = scene;
    play->sceneId = sceneId;
    play->sceneConfig = scene->drawConfig;
    std::string scenePath =
        StringHelper::Sprintf("scenes/nonmq/%s/%s", scene->segment.fileName, scene->segment.fileName);
    play->sceneSegment = OTRPlay_LoadFile(play, scenePath.c_str());
    scene->unk_D = 0;
    gSegments[2] = (uintptr_t)play->sceneSegment;
    OTRPlay_InitScene(play, spawn);
    Room_SetupFirstRoom(play, &play->roomCtx);
    
    GameInteractor_Should(VB_RANDOMIZE_ALL_ON_NEW_SCENE, false);  
}
``` 

It does work and the audio is randomized when moving to different areas in clock town  
`South Clock Town -> North Clock Town`

Eblo replied:
When you call `COND_VB_SHOULD(VB_SOMETHING, someCondition)`, the code within will not run unless `someCondition` is true, so it's redundant to check `someCondition` inside the macro.

I checked out your changes locally to do some digging and found that my suggestion still won't work in areas like Clock Town anyway, like you've said, and the reason is that transitions within Clock Town do not reach that hook. When I try OnSceneInit, I believe it occurs too late, as I get no BGM with a display that shows the vanilla BGM name.

Then I tried `VB_PLAY_TRANSITION_CS`, which occurs on scene transitions before the actual scene load and seems to work as you intended. If no one sees any other issues with this approach, I think you can do this:

```c++
void AudioEditor::InitElement() {
    COND_VB_SHOULD(VB_PLAY_TRANSITION_CS, CVAR_AUDIO("RandomizeAllOnNewScene"), {
        AudioEditor_RandomizeAll();
    });

    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnRandoSeedGeneration>([]() {
        if (CVarGetInteger(CVAR_AUDIO("RandomizeAllOnRandoGen"), 0)) {
            AudioEditor_RandomizeAll();
        }
    });
}
```

Note that I condensed the various functions into one since there didn't seem to be any need to have them outside of `InitElement`.

Glought replied:
The reason I had a if statement inside `COND_VB_SHOULD` is because it ignored `CVAR_AUDIO("RandomizeAllOnRandoGen")` and it would trigger the code inside the macro no matter if `"Randomize All Music and Sound Effects on New Scene"` was checked or unchecked it all ways triggered. 

It didn't matter if the ` GameInteractor_Should(VB_RANDOMIZE_ALL_ON_NEW_SCENE, false);  ` was set to true or false.
Maybe being triggered from  `z_play_2SH.cpp` in the `OTRPlay_SpawnScene` function has something to do about it ignoring the condition.

Glought replied:
I just tested `VB_PLAY_TRANSITION_CS` and it still ignores the condition and triggers the code inside the macro.
Also the music cuts out when transitioning to the different areas of clock town after loading a game save but then works fine after entering and leaving a shop or clock town then the music works like it should.

Glought replied:
I forgot about `static RegisterShipInitFunc initFunc` so i placed 

```c++
static RegisterShipInitFunc initFuncRandomizeAllOnNewScene(AudioEditorRegisterRandomizeAllOnNewScene,
                                     { CVAR_AUDIO("RandomizeAllOnNewScene") });
``` 
at the end of AudioEditor and now it follows the condition.

Eblo replied:
Yep, that would explain it. Without using ShipInit before, it would only evaluate the condition once and not refresh on changes.

Glought replied:
Any suggestions on the Music stops playing on first transition after game load issue?.

Eblo replied:
I think you should be using `VB_PLAY_TRANSITION_CS`:

```c++
void AudioEditorRegisterRandomizeAllOnNewScene() {
    COND_VB_SHOULD(VB_PLAY_TRANSITION_CS, CVAR_AUDIO("RandomizeAllOnNewScene"), { AudioEditor_RandomizeAll(); });
}

static RegisterShipInitFunc initFuncRandomizeAllOnNewScene(AudioEditorRegisterRandomizeAllOnNewScene,
                                                           { CVAR_AUDIO("RandomizeAllOnNewScene") });
```

`VB_PLAY_TRANSITION_CS` occurs just before the game loads a new scene. I think your `VB_RANDOMIZE_ALL_ON_NEW_SCENE` occurs too late, `OnSceneInit` occurs too late, and `VB_PLAY_SCENE_SEQUENCE` does not run between scenes with shared music like Clock Town.

Glought replied:
I am using `VB_PLAY_TRANSITION_CS` during my tests today.  
Here I recorded the issue




Music stops on first transition after game load.
https://github.com/user-attachments/assets/5e6b8c97-14b7-442a-9ba5-6fc78ce139bc

Music works fine after entering and leaving a shop.
https://github.com/user-attachments/assets/1cc52b17-25fd-4cb0-9720-62aca1748235

Eblo replied:
Go ahead and push what you have, even if it's not fully functional. Locally, I have your PR as-is but with the previous code snippet I shared, and I'm not getting that behavior when loading in South Clock Town and going west.

Glought replied:
ok i pushed what i got

Eblo replied:
Would you mind sanity checking the PR build? It works as intended on my end where I boot up, load into a game, appear in South Clock Town, and go to West Clock Town. With how this only rebuilds in pieces, it's possible something went stale and caused the issue for you locally. If you can confirm the PR build is working, then I'd try deleting your build folder and rebuilding from scratch to try again.

Glought replied:
Deleting the build folder and rebuilding fixed the issue. I thought about doing that yesterday but didn't have the time till now.

Eblo commented on mm/src/code/z_actor.c:
```diff
@@ -2486,8 +2486,12 @@ void Player_PlaySfx(Player* player, u16 sfxId) {
+        } else {
+            AudioSfx_PlaySfx(sfxId, &player->actor.projectedPos, 4, &gSfxDefaultFreqAndVolScale,
+                             &gSfxDefaultFreqAndVolScale, &gSfxDefaultReverb);
+        }
```
Let's try not to change flow if we can help it. I know there are areas that do this, but let's try and keep that to a minimum, only doing so if necessary.

```suggestion
        if (GameInteractor_Should(VB_LINK_VOICE_PITCH_MULTIPLIER, true, &sfxId)) {
            AudioSfx_PlaySfx(sfxId, &player->actor.projectedPos, 4, &gSfxDefaultFreqAndVolScale,
                             &gSfxDefaultFreqAndVolScale, &gSfxDefaultReverb);
        }
```

Eblo commented on mm/2s2h/Enhancements/Audio/Sfx/LinksVoicePitchMultiplier.cpp:
```diff
@@ -0,0 +1,37 @@
+
+void RegisterLinksVoicePitchMultiplier() {
+    COND_VB_SHOULD(VB_LINK_VOICE_PITCH_MULTIPLIER, CVAR, {
+        *should = true;
```
Paired with my suggested flow change, this would need to be set to false later, after checking the condition.

```suggestion
```

Eblo commented on mm/2s2h/Enhancements/Audio/Sfx/LinksVoicePitchMultiplier.cpp:
```diff
@@ -0,0 +1,37 @@
+        } else {
+            AudioSfx_PlaySfx(sfxId, &player->actor.projectedPos, 4, &gSfxDefaultFreqAndVolScale,
+                             &gSfxDefaultFreqAndVolScale, &gSfxDefaultReverb);
+        }
```
With how `should` works, you won't have to re-implement default behavior. It will just happen by default if you don't set `should` to false.

```suggestion
        }
```

Eblo commented on mm/2s2h/Enhancements/Audio/Sfx/LinksVoicePitchMultiplier.cpp:
```diff
@@ -0,0 +1,37 @@
+        u16 sfxId = *va_arg(args, u16*);
+
+        if (sfxId >= NA_SE_VO_LI_SWORD_N && sfxId <= NA_SE_VO_DEMO_394 || sfxId == NA_SE_PL_TRANSFORM_VOICE) {
+
```
At this point in your code within the conditional, we know we're processing a Link voice sfx, so we know not to do the vanilla behavior, thus we can set this here.
```suggestion
            *should = false;
```

Eblo reviewed (changes requested):
Appreciate the cleanup and organizational changes! I have some tidy up suggestions. Randomizing audio with `OnRoomInit` is an issue, as I don't think it has the intended effect you were going for. It might seem like the answer in most scenes, but scenes with multiple rooms, like dungeons, will result in somewhat frequent audio shuffling.

Glought committed:
Made Changes requested by Eblo to "LinksVoicePitchMultiplier.cpp" and
"z_actor.c"

Glought committed:
WIP Randomize On New Scene
